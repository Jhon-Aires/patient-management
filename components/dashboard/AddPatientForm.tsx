'use client';

import { zodResolver } from '@hookform/resolvers/zod';
import { SubmitHandler, useForm } from 'react-hook-form';
import { z } from 'zod';

import { Button } from '@/components/ui/button';
import { DialogFooter } from '@/components/ui/dialog';
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage
} from '@/components/ui/form';
import { formFields, initialValues, schema } from '@/lib/forms/addPatientForm';
import { Patient } from '@/types/Patients';

export default function AddPatientForm({
  onSave,
  onCancel
}: {
  onSave: (patient: Patient) => void;
  onCancel: () => void;
}) {
  const form = useForm<z.infer<typeof schema>>({
    resolver: zodResolver(schema),
    defaultValues: { ...initialValues },
    reValidateMode: 'onSubmit'
  });

  const onSubmit: SubmitHandler<typeof initialValues> = ({
    avatar,
    description,
    name
  }) => {
    onSave({
      avatar,
      description,
      name,
      id: 1,
      createdAt: new Date().toISOString(),
      website: ''
    });
  };

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-4">
        {formFields.map(
          ({
            controller: Controller,
            controllerAttrs,
            label,
            name,
            description
          }) => {
            console.log('controllerAttrs', controllerAttrs);

            return (
              <FormField
                key={name}
                control={form.control}
                name={name as keyof typeof initialValues}
                render={({ field }) => {
                  console.log('controllerAttrs', controllerAttrs);

                  return (
                    <FormItem>
                      {label && <FormLabel>{label}</FormLabel>}
                      <FormControl>
                        <Controller
                          {...field}
                          {...controllerAttrs}
                          className={controllerAttrs?.className}
                        />
                      </FormControl>
                      {description && (
                        <FormDescription>{description}</FormDescription>
                      )}
                      <FormMessage />
                    </FormItem>
                  );
                }}
              />
            );
          }
        )}
        <DialogFooter>
          <Button type="button" variant="outline" onClick={onCancel}>
            Cancel
          </Button>
          <Button type="submit">Add</Button>
        </DialogFooter>
      </form>
    </Form>
  );
}
