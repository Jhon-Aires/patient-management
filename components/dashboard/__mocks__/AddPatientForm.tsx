export default function AddPatientForm() {
  return <div>AddPatientForm</div>;
}
