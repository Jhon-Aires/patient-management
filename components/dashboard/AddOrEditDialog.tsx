import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle
} from '@/components/ui/dialog';
import EditPatientForm from '@/components/dashboard/EditPatientForm';
import { Patient } from '@/types/Patients';
import AddPatientForm from './AddPatientForm';

export default function AddOrEditDialog({
  editingPatient,
  handleCancel,
  handleSave,
  isAddingPatient
}: {
  editingPatient: Patient | null;
  handleCancel: () => void;
  handleSave: (patient: Patient) => void;
  isAddingPatient: boolean;
}) {
  return (
    <Dialog open={editingPatient !== null} onOpenChange={handleCancel}>
      <DialogContent>
        <DialogHeader>
          <DialogTitle>
            {isAddingPatient ? 'Add Patient' : 'Edit Patient'}
          </DialogTitle>
          <DialogDescription>
            {isAddingPatient
              ? 'Add a new patient to the system.'
              : 'Make changes to patient information here.'}
          </DialogDescription>
        </DialogHeader>
        {editingPatient && !isAddingPatient && (
          <EditPatientForm
            patient={editingPatient}
            onSave={handleSave}
            onCancel={handleCancel}
          />
        )}
        {editingPatient && isAddingPatient && (
          <AddPatientForm onSave={handleSave} onCancel={handleCancel} />
        )}
      </DialogContent>
    </Dialog>
  );
}
