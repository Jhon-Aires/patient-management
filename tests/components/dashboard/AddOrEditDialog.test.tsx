import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import AddPatientForm from '@/components/dashboard/AddPatientForm';
import AddOrEditDialog from '@/components/dashboard/AddOrEditDialog';

jest.mock('@/components/dashboard/AddPatientForm');
jest.mock('@/components/dashboard/EditPatientForm');

describe('AddOrEditDialog', () => {
  const editingPatient = {
    avatar: 'patient avatar',
    createdAt: '',
    description: 'description',
    id: 1,
    name: 'Patient Name',
    website: ''
  };

  const onCancel = jest.fn();
  const onSave = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { asFragment } = render(
      <AddOrEditDialog
        editingPatient={editingPatient}
        handleCancel={onCancel}
        handleSave={onSave}
        isAddingPatient={false}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should show the EditPatientForm when editing a patient', () => {
    render(
      <AddOrEditDialog
        editingPatient={editingPatient}
        handleCancel={onCancel}
        handleSave={onSave}
        isAddingPatient={false}
      />
    );

    expect(screen.getByText(/editpatientform/i)).toBeInTheDocument();
  });

  it('should show the AddPatientForm when adding a patient', () => {
    render(
      <AddOrEditDialog
        editingPatient={editingPatient}
        handleCancel={onCancel}
        handleSave={onSave}
        isAddingPatient={true}
      />
    );

    expect(screen.getByText(/addpatientform/i)).toBeInTheDocument();
  });
});
