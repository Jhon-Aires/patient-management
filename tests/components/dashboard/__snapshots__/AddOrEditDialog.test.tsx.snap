// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddOrEditDialog should render correctly 1`] = `<DocumentFragment />`;
